changepass.man
.\" changepass 1.4.0preX
.\"
.TH CHANGEPASS 8 "DEC ??, 2015"
.AT 8
.SH NAME
    \fBchangepass\fP \- update an user's password

.SH SYNOPSIS
    changepass
    
.SH DESCRIPTION
\fBchangepass\fP is a chpasswd clone, it might be useful on 
platforms that doesn't have such a command like Solaris.
.br
.br
Most GNU/Linux distributions have chpasswd (8), on FreeBSD 
you can use "pw usermod name -h 0" but many commercial Un*ces
doesn't have tool like this.  An alternative is to update the 
user's password in script with \fBusermod\fP
.br
.br
but it's possible to see the encrypted password in the process
list, which is not very secure.
.br
.br
\fBchangepass\fP reads a list of user name and password pairs
from \fBstdin\fP  and updates the users passwords
.br
.br
Each line has format:
.br
username:password
.br    
.SH OPTIONS
.br
 -h,--help              print this help
 -n,--nopam             don't use pam
 -p,--pam               use pam (default)
 -e,--encrypt           password is already encrypted, this option
.br
                        will disable pam
 -m,--md5               use md5 algorithm, this option will
.br
                        disable pam
 -c,--crypt-method      use des|md5|sha256|sha512 crypt method this
.br
                        option will disable pam
 -s,--sha-rounds        number of sha2 rounds
 -v,--verbose           enable verbose output
.br    
.SH USAGE
.br 
cat user_password_pairs | changepass
.br

.SH AUTHOR
Written by Staf Wagemakers <staf@wagemakers.be>
    
.SH COPYRIGHT
.SS Documentation
The documentation for CGIpaf is licensed under the terms of
the FreeBSD Documentation License.
.PP
The FreeBSD Documentation License

Copyright (c) 2001 - 2020 Staf Wagemakers

Redistribution and use in source (plaintext, HTML) and 'compiled' forms
(SGML, HTML, PDF, PostScript, RTF and so forth) with or without
modification, are permitted provided that the following conditions are
met:

1. Redistributions of source code (plaintext, HTML) must retain the
above copyright notice, this list of conditions and the following
disclaimer.

2. Redistributions in compiled form (transformed to other DTDs,
converted to PDF, PostScript, RTF and other formats) must reproduce
the above copyright notice, this list of conditions and the following
disclaimer in the documentation and/or other materials provided with
the distribution.

THIS DOCUMENTATION IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED 
WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF 
MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. 
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS DOCUMENTATION,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.SS Software
.br
CGIpaf (the software) is licensed under the terms of the
GNU General Public License version 2 or later.
.PP
Copyright (c) 2002, 2007 Staf Wagemakers.
.PP
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
.br


ChangeLog
* version 1.3.5		( 13 Apr 2016)
  + BUGFIX: pwd_mkdb errors were not handled correctly on BSD 
  + kyua test scripts added 

* version 1.3.4		( 23 Nov 2014 )
  + Cracklib configuration checking has been improved
  + LDFLAGS is passed to the linker correctly

* version 1.3.4pre1	( 15 Sep 2013 )
  + Enable PAM on FreeBSD 7.3 or above  
  + Enable PAM on NetBSD 6.0 or above
  + xmalloc updated to support systems with non GNU compatible malloc

* version 1.3.3		( 24 Jan 2008 )
  + BUGFIX: md5 passwords on BIG ENDIAN systems were not calculated correctly

* version 1.3.2		( 30 Aug 2007 )
  + changepass improved
  + added formail directive
  + BUGFIX: lock dir is detected correctly
  + BUGFIX: corrected compile warnings
  + BUGFIX: vacation_cgi example has been update to work with newer
            versions of perl/cgi module

* version 1.3.1		( 13 jul 2004 )
  + BUGFIX: install.sh will now ask to create the cgi-bin directory
            (Thanks to Koen de Jong for reporting this problem)
  + BUGFIX: configure --disable-pam works now on PAM systems
  + BUGFIX: fixed compile problem on non-PAM systems
            (Thanks to Peter Hübert for reporting this problem)
  + Dutch translation added  (Thanks to Koen de Jonge)   

* version 1.3.0	        ( 26 Apr 2004 )
  + Documentation updated

* version 1.3.0rc2	( 10 Mar 2004 )
  + PAM: the pam functions should now work on Solaris 9+ 

* version 1.3.0rc1	( 6 Jan 2004 )
  + autoconf has been upgraded to 2.57: the build host is auto-detected
  + new variables:
        _POST[] array : An associative array of variables with the original
	                HTTP POST
        post_string   : A string with the original HTTP POST
	message       : A string with the default message
  + new directives to enable more advanced mail configuration scripts:
    run_viewmailcfg, mailcfg_check. a sample perl vacation script is included.
  + set_script_filename and unset_script_filename directives 
    ( useful if you want to use a php script )

* version 1.3.0pre4    	( 30 May 2003 )
  + new forward method to avoid mail looping

* version 1.3.0pre3    	( 28 Apr 2003 )

  + BUGFIX: setuid(0) added before the passwd run_success script, to
    re-enable smbpasswd support. which was broken in the previous release.

* version 1.3.0pre2    	( 31 Mar 2003 )
  
  + NetBSD support
  + BUGFIX: 1.3.0pre1 didn't compile on FreeBSD
  
  + merged from CGIpaf 1.2.1
   
     - example in Brazilian Portuguese added ( Thanks to Fabricio Machado )
     - improved forward_to emailaddress testing
     - added test for mail headers like 'Precedence: bulk' to avoid replying 
       to maillinglists etc.
     - code cleanup

* version 1.3.0pre1     (  2 Jan 2003 )
  + linuxconf virtual mail password files support added

* version 1.2.0         (  7 Dec 2002)
  + 1.2.0RC1 has been released as 1.2.0 

* version 1.2.0RC1      ( 26 Nov 2002 )
  + setreuid(user_uid,0) added in cgipaf.c 
    passwd.cgi should now support all pam options in the PAM configuration 
    (cracklib, minimum/ maximum password length etc.) on all systems with PAM
    support.

  + The set_PAM_CHANGE_EXPIRED_AUTHTOK directive is obsolete and will be 
    removed in the next releases of CGIpaf

* version 1.2.0pre1     ( 24 Oct 2002 )
  + Access Control List added ( AclOrder, AllowUsers, AllowGroups, DenyUsers
    DenyGroups directives )

* version 1.1.0         ( 17 Sep 2002 )
  + Documentation update

* version 1.1.0rc2      (  07 Aug 2002 )
  + The ability to list words in /etc/cgipasswd that should not be a part of
    new passwords  ( Thanks to Daniel R for his patch )

* version 1.1.0rc2pre1  (  24 Jul 2002 )
  + FIXED: run_after_mailcfg was executed before the mail configuration

* version 1.1.0rc1      (  15 Jul 2002 )
  + changepass utility added

* version 1.1.0pre2      ( 26 Jun 2002 )
  + new run_mailcfg directive added to support other mail configurations than 
    procmail a sample script with vacation support is included.

* version 1.1.0pre1      ( 18 Jun 2002 )
  + new run_before_mailcfg and run_after_mailcfg directive
  + new use_statefile directive

* version 1.0.1          (  4 Jun 2002 )
  + improved cracklib detection

* version 1.0.1rc1       ( 21 May 2002 )
  + updates to configure
  + documentation update

* version 1.0.1rc1pre2   ( 17 May 2002 ) 
  + FreeBSD support

* version 1.0.1rc1pre1   ( 29 Apr 2002 )
  + md5 password support on non-pam systems ( Slackware )

* version 1.0.0          ( 23 Apr 2002 )
  + faq
  + documentation update
  + new install script

* version 1.0.0pre1.2    ( 3 Apr 2002 )
  + more logging in debug mode
  + documentation update

* version 1.0.0pre1.1    ( 28 Mar 2002 )
  + new password unchanged directive
  + updates to pass.c to support more non-GNU/Linux Unices

* version 1.0.0pre1.0    ( 19 Mar 2002 )
  + new cracklib_dictpath directive
  + FIXED cgipaf did't compile on systems where CRACKLIB_DICTPATH wasn't 
    defined
  + update to configure

* version 1.0rc5         ( 28 Jan 2002 )
  + documentation update
  + php files update

* version 1.0rc5pre2     ( 16 Jan 2002 )
  + new pam conversation function
  + new set_PAM_CHANGE_EXPIRED_AUTHOK directive

* version 1.0rc5pre1     ( 20 Dec 2001 )
  + cracklib support
  + support for all common pam messages ( cracklib, minimum password length etc )

* version 1.0rc4     ( 4 Dec 2001 )
  + documentation update

* version 1.0rc4pre5 ( 28 Nov 2001 )
  + new pam_service directive, you can change the pam service name.
  + added test for gdbm/ndbm.h header file
  + fixed: mailcfg.cgi failed, if it couldn't determine the domain name.
  
* version 1.0rc4pre4 ( 11 Nov 2001 )
  + new forwarding method to avoid mail looping
  + fixed: it wasn't possible to use special characters ( $ space ) as the
    loginname / password.

* version 1.0rc4pre3 ( 27 oct 2001)
  + fixed: run_success, run_locked didn't always work due a buffer overflow in
    xstring.c
  + fixed: typo the sample samba passwd script the configuration file & 

changepass.c
/*
 * changepass.c
 *
 * Copyright (C) 2002, 2006, 2007, 2015, 2016, 2020 Staf Wagemakers Belgie/Belgium
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 */

#include "common.h"
#include "pass.h"
#include <libgen.h>
#include <errno.h>


char *prgname;

int isoptarg(char *optstr) {

	if(optstr==NULL) return 0;

	switch(*optstr) {

		case 0:
				return(0);
				break;
		case '-':	
				return(0);
				break;
		default:
				return(1);
				break;

	}

	return(0);

}

/*
 * display usage message
 */

void usage() {

	char txt_copyright[]="                        (GPL) 2002-2016 Belgie  Belgium\n                                                            Staf Wagemakers\n                                                  email: staf@wagemakers.be";

	char txt_usage000[]="[OPTION]\n\noptions:\n\n -h,--help\t\tprint this help\n -n,--nopam\t\tdon't use pam\n";
	char txt_usage001[]=" -p,--pam\t\tuse pam (default)\n";
	char txt_usage002[]=" -e,--encrypt\t\tpassword is already encrypted";
	char txt_md5_supported[]=" -m,--md5\t\tuse md5 algorithm";
	char txt_md5_unsupported[]=" -m,--md5\t\tuse md5 algorithm (unsupported on this platform)";
	char *txt_usage003;

	if(is_crypt_supported("md5")) {

		txt_usage003=txt_md5_supported;

	}
	else {
		txt_usage003=txt_md5_unsupported;


	}

	char txt_usage004[]=" -c,--crypt-method\tuse";
	char txt_usage005[]="crypt method";
	char txt_usage006[]=" -s,--sha-rounds\tnumber of sha2 rounds\n";
	char txt_usage007[]=" -v,--verbose\t\tenable verbose output\n\n";

#ifdef _WITHPAM
	char txt_disable_pam[]=", this option will disable pam\n";
#else
	char txt_disable_pam[]="\n";
#endif

	char * txt_supported_algo=combine_strings(xcrypt_supported_crypts());
	replace_char(txt_supported_algo,' ','|');

      	fprintf(stderr,"%s %s",prgname,(char *)CGIPAF_VERSION);
	fprintf(stderr,"%s\n\n",txt_copyright);


      	fprintf(stderr,"usage: %s %s",prgname,txt_usage000);
#ifdef _WITHPAM
	fprintf(stderr,"%s",txt_usage001);
#endif
	fprintf(stderr,"%s%s",txt_usage002,txt_disable_pam);
#ifdef MD5_CRYPT
	fprintf(stderr,"%s%s",txt_usage003,txt_disable_pam);
#endif
 	fprintf(stderr,"%s %s %s\n\t\t\t%s",txt_usage004,txt_supported_algo,txt_usage005,txt_disable_pam);
#ifdef MODERNCRYPT_SHA2 
 	fprintf(stderr,"%s",txt_usage006);
#endif
 	fprintf(stderr,"%s",txt_usage007);
	fprintf(stderr,"\n");
}

struct  pw_info * get_pw_info (char *name,int pamflag) {

   	struct pw_info *pw;

	#ifdef _WITHPAM

   	if (pamflag) {

	   	pw=get_pw(name);

   	}
   	else {
	   	pw=get_pw_nopam(name);

   	}

	#else

   		pw=get_pw(name);

	#endif

	return(pw);

}


int main (int argn,char **argv) 
{
   int i;
   char line[1024]="\0";
   char *pass;
   char **passwords=NULL;
   char *name;
   char **names=NULL;
   char *c;
   struct pw_info *pw;
   int hlpflag=0;
   int encryptflag=0;
   int c_nopamflag=0;
   int nopamflag=0;
   int nopammode=-1;
   int pamflag=0;
   int md5flag=0;
   int algoflag=0;
   char *algorithme=NULL;
   char algomode=-1; 
   int sharoundsflag=0;
   char *sharounds_str=NULL;
   int sharounds=0;
   int verboseflag=0;
   int linecounter=0;
   int number_of_lines=0;
   int error=0;
   int pass_updated=0;
   
   prgname=basename(argv[0]);              /* set prgname to the real program name */

   if(argn>1) {                  /* we've no arguments */

	   int i;
	   char *longargs[]={"encrypted","md5","help","nopam","pam","verbose","crypt-method","--sha-rounds",NULL};

	   /*
	    * convert long opts to shorts opt
	    */

	   for (i=1;i<argn;i++) {

		   char *shortarg=NULL;

		   if (strncmp("--",argv[i],2)==0) {

			   	char **ccp;
				int found=0;

			   	for(ccp=longargs;*ccp!=NULL;++ccp) {

					if (!strcmp(argv[i]+2,*ccp)) {
			   	
						shortarg=xmalloc(3*sizeof(char));
						shortarg[0]=0;
						strncat(shortarg,(argv[i]+1),2);
						argv[i]=shortarg;
						found=1;
						break;

					}

				}

				if (!found) {
						fprintf(stderr,"%s: Unkown argument %s\n",prgname,argv[i]);
				        	hlpflag=1;

				}
		   }
	   }

	   if(!hlpflag) {

	   	while ((i = getopt(argn, argv, "hpemnvc:s:")) != -1) {

			switch(i) {

		   		case 'h':
						hlpflag=1;
						break;
		   		case 'e':
						if (encryptflag) {

							fprintf(stderr,"Sorry, -e can only be used ones\n\n");
							hlpflag=1;

						}

						encryptflag=1;
						nopammode=254;
						break;
		   		case 'm':
						if (md5flag) {

							fprintf(stderr,"Sorry, -m can only be used ones\n\n");
							hlpflag=1;

						}

						if (!is_crypt_supported("md5")) {

							fprintf(stderr,"Sorry, md5 is not supported on this platform\n\n");
							hlpflag=1;

						}

						md5flag=1;
						nopammode=1;
						break;
		   		case 'n':
						if (c_nopamflag) {

							fprintf(stderr,"Sorry, -n can only be used ones\n\n");
							hlpflag=1;

						}

						c_nopamflag=1;
						nopamflag=1;
						break;
		   		case 'p':
						if (pamflag) {

							fprintf(stderr,"Sorry, -p can only be used ones\n\n");
							hlpflag=1;

						}

						pamflag=1;
						break;
				case 'v':
						if (verboseflag) {

							fprintf(stderr,"Sorry, -v can only be used ones\n\n");
							hlpflag=1;

						}

						verboseflag=1;
						break;
		   		case '?':	
						hlpflag=1;
						break;
				case 'c':
						if (algoflag) {

							fprintf(stderr,"Sorry, -c can only be used ones\n\n");
							hlpflag=1;

						}

						algoflag=1;

						if(!isoptarg(optarg)) {

							fprintf(stderr,"Sorry, option -c requires an argument.\n\n");
							hlpflag=1;
							break;

						}

						algorithme=optarg;

						algomode=cryptstr2int(algorithme);

						if(algomode==-1) {
							fprintf(stderr,"Sorry, unknown crypt type (%s).\n\n",null2str(algorithme));
							hlpflag=1;

						} else {

							if (is_crypt_id_supported(algomode)) {

								nopammode=algomode;

							} else {
								fprintf(stderr,"Sorry, unsupported crypt type name = (%s) crypt_id = (%d).\n\n",null2str(algorithme),algomode);
								hlpflag=1;


							}			

						}

						break;
				case 's':
						if (sharoundsflag) {

							fprintf(stderr,"Sorry, -s can only be used ones\n\n");
							hlpflag=1;

						}

   						sharoundsflag=1;

						if(!isoptarg(optarg)) {

							fprintf(stderr,"Sorry, option -s requires an argument.\n\n");
							hlpflag=1;
							break;

						}

						sharounds_str=optarg;

						if(!isstrdigit(sharounds_str)) {

							fprintf(stderr,"Sorry, option -s needs to be a digit string\n\n");
							hlpflag=1;
							break;

						}

						int i=sscanf(sharounds_str,"%d",&sharounds);

						if(i!=1) {

	   						fprintf(stderr,"sscanf failed (i=%d) is not 1, errno=%d errstr=%s",i,errno,null2str(strerror(errno)));
							exit(3);

						}

						char sharounds_test_buffer[1024];

						if (snprintf(sharounds_test_buffer,1023,"%d",sharounds)>=1023) {

							fprintf(stderr,"Invalid nummeric value length>1023\n");
							hlpflag=1;
							break;

						}

						if(strcmp(sharounds_test_buffer,sharounds_str)) {

							fprintf(stderr,"Invalid nummeric value\n");
							hlpflag=1;
							break;

						}

						break;

				default:
						hlpflag=1;
		   	}

	   	}

	   }

	   if ( verboseflag ) {
		
		fprintf(stderr,"DEBUG: option flags\n\n");

	   	fprintf(stderr,"hlpflag = %d\n",hlpflag);
	   	fprintf(stderr,"encryptflag = %d\n",encryptflag);
	   	fprintf(stderr,"pamflag = %d\n",pamflag);
	   	fprintf(stderr,"nopamflag = %d\n",nopamflag);
	   	fprintf(stderr,"nopammode = %d\n",nopammode);
	   	fprintf(stderr,"md5flag = %d\n",md5flag);
	   	fprintf(stderr,"algoflag = %d\n",algoflag);
	   	fprintf(stderr,"algorithme = %s\n",null2str(algorithme));
	   	fprintf(stderr,"sharoundsflag = %d\n",sharoundsflag);
	   	fprintf(stderr,"sharounds_str = %s\n",null2str(sharounds_str));
	   	fprintf(stderr,"sharounds = %d\n",sharounds);
	   	fprintf(stderr,"verboseflag = %d\n",verboseflag);

	   }

	
	/*
 	 * invalid arguments handling
	 */

#ifndef _WITHPAM

	if(pamflag) {

		fprintf(stderr,"%s: Sorry, this version is compiled without pam support.\n\n",prgname);
	   	hlpflag=1;

	}

#endif
	   if (pamflag && nopamflag) {

		   fprintf(stderr,"%s: Can't enable pam and nopam.\n\n",prgname);
		   hlpflag=1;

	   }
	   else {

		   if ( pamflag && ( md5flag || encryptflag || algoflag ) ) {

			   fprintf(stderr,"%s: --md5, --crypt-method and --encrypted are invalid with forced pam.\n\n",prgname);

			   hlpflag=1;

		   }
		   else {
		   	if ( md5flag || encryptflag || algoflag ) {

				nopamflag=1;

				if (md5flag && encryptflag) {

					fprintf(stderr,"%s: Can't use --md5 and --encrypted at the same time\n\n",prgname);
					hlpflag=1;

				}

				if (md5flag && algoflag) {

					fprintf(stderr,"%s: Can't use --md5 and --crypt-method at the same time\n\n",prgname);
					hlpflag=1;

				}

				if (algoflag && encryptflag) {

					fprintf(stderr,"%s: Can't use --crypt-method and --encrypted at the same time\n\n",prgname);
					hlpflag=1;

				}

			} else {

				if (sharoundsflag) {

					if (!algoflag  ) {

						fprintf(stderr,"%s: Sorry --sha-rounds is only allowed with a sha2 algorithm\n\n",prgname);
						hlpflag=1;

					} else {

						switch (algomode) {

							case 3:
							case 4:
								break;
							default:
								fprintf(stderr,"%s: 2 Sorry --sha-rounds is only allowed with a sha2 algorithm\n\n",prgname);
								hlpflag=1;
								break;


						}
					}

				}

			}

		   }

	   }

	   /*
 	    *  display usage and exit if hlpflag is enabled.
 	    */

	   if(hlpflag) {

	   	usage();
	   	exit(1);

	   }

   }
   

#ifdef _WITHPAM

  /*
   * enable pam by default on pam systems
   */

  if (!nopamflag) {

	   pamflag=1;

  }

#ifdef NO_CHPASS_CRYPT

  /* 
   * on pam only system nopam is not available
   */

  else {

	  fprintf(stderr,"%s: is compiled without --nopam support\n\n",prgname);
	  exit(1);

  }

#endif
   
   set_pam_service("passwd");      /* set PAM service name */
   set_pam_chauth_flag(0);

#else

        /*
         * On non-pam systems the --pam argument is invalid
         */

	if (pamflag) {

		fprintf(stderr,"%s: is compiled without --pam support\n\n",prgname);
		exit(1);


	}
	else {

		pamflag=0;
		nopamflag=1;

	}
   
#endif

   	if ( verboseflag ) {

		fprintf(stderr,"DEBUG: option flags after evaluation\n\n");

   		fprintf(stderr,"hlpflag = %d\n",hlpflag);
   		fprintf(stderr,"encryptflag = %d\n",encryptflag);
   		fprintf(stderr,"pamflag = %d\n",pamflag);
   		fprintf(stderr,"nopamflag = %d\n",nopamflag);
   		fprintf(stderr,"md5flag = %d\n",md5flag);
	   	fprintf(stderr,"algoflag = %d\n",algoflag);
	   	fprintf(stderr,"algorithme = %s\n",null2str(algorithme));
	   	fprintf(stderr,"sharoundsflag = %d\n",sharoundsflag);
	   	fprintf(stderr,"sharounds_str = %s\n",null2str(sharounds_str));
	   	fprintf(stderr,"sharounds = %d\n",sharounds);
   		fprintf(stderr,"verboseflag = %d\n",verboseflag);

   	}

	switch(algomode) {

		case 3:
		case 4:
			if(sharoundsflag) {

				if (verboseflag) {

					fprintf(stderr,"DEBUG: set sha2 rounds to %d\n",sharounds);

				}

				set_sha2_prefered_rounds(sharounds);

				if(get_sha2_prefered_rounds() != sharounds ) {

					fprintf(stderr,"ERROR: set sha2rounds failed\n");
					exit(3);


				}

			}

	}


	/*
	 * parse stdin line by line
	 */

	linecounter=0;

	for (;;) {

   		if (fgets(line,1023,stdin)==NULL) break;          /* read the data from stdin */

   			if(strlen(line)==0) {

				if (linecounter == 0 ) {

      					usage();                      /* oops, no data */
      					exit(1);

				}else {
					continue;
				}

   		}

		/* 
		 * detect a buffer overrun
		 */
   
   		if(line[strlen(line)-1]!='\n') {

      			fprintf(stderr,"\n%s: Too much data...\n\n",prgname);  /* buffer too small */
      			exit(1);

   		}


		/*
		 * chop
		 */

		if (line[strlen(line)-1] == '\n' ) {

   			line[strlen(line)-1]='\0'; 

		}

		/*
		 * Ignore empty lines
		 */

		if (strlen(line)==0) {

			continue;

		}

		/*
		 * check on password seperator
		 */
   
   		if((c=strstr(line,":"))==NULL) {

      			fprintf(stderr,"\n%s: Input error no \":\" in input on line: %d\n\n",prgname,linecounter+1);
      			exit(1);

   		}
   
   		/* copy the username into name */
   
   		name=(char *) xmalloc((c-line)+1);
   		strncpy(name,line,(c-line));
   		name[c-line]='\0';
		names=(char **) xrealloc(names,(linecounter+2)*sizeof(char **));
		names[linecounter]=name;
		names[linecounter+1]=NULL;
   
   		/* copy the password into pass */
   
   		pass=(char *) xmalloc(strlen(c)+1);
   		/* strncpy(pass,c+1,strlen(c)-1); */
   		strcpy(pass,c+1);
   		/* pass[strlen(c)]='\0'; */
		passwords=(char **) xrealloc(passwords,(linecounter+2)*sizeof(char **));
		passwords[linecounter]=pass;
		passwords[linecounter+1]=NULL;

		++linecounter;

	}

	number_of_lines=linecounter;

	/*
	 * loop through the user and vrfy that we can get the password info
	 */

	for (linecounter=0;linecounter<number_of_lines;++linecounter) {

/*
		if (verboseflag) fprintf(stderr,"DEBUG \"%s\" \"%s\"\n",names[linecounter],passwords[linecounter]);

*/

		name=names[linecounter];
		pass=passwords[linecounter];

		pw=get_pw_info(name,pamflag);


   		if (!pw) {

      			fprintf(stderr,"\n%s: get_pw() failed. Failed to get password info for user: %s on line: %d\n",prgname,name,linecounter+1); /* get_pw() failed */
			fprintf(stderr,"%s: Errors detected, changes ignored\n\n",prgname);
      			exit(1);

   		};

	}

	/*
	 * walk again to update the passwords
	 */

	for (linecounter=0;linecounter<number_of_lines;++linecounter) {

/*
		if (verboseflag) fprintf(stderr,"DEBUG \"%s\" \"%s\"\n",names[linecounter],passwords[linecounter]);
*/


		name=names[linecounter];
		pass=passwords[linecounter];

		pw=get_pw_info(name,pamflag);


   		if (!pw) {

      			fprintf(stderr,"\n%s: Failed to get password info for user: %s on line: %d\n",prgname,name,linecounter+1); /* get_pw() failed */
			error=1;
      			continue;

   		};

#ifdef _WITHPAM

   		if (pamflag) {

			if (verboseflag) fprintf(stderr,"DEBUG: running chpw() for user \"%s\"\n",name);
			i=chpw(pw,pass);

   		}
   		else {

#endif

	   		int mode=255;

			if (algoflag) {

				mode=algomode;

			} else {

	   			if (md5flag) {

		   			mode=1;


	   			}
	   			else {

		   			if(encryptflag) {

			   			mode=254;

		   			}
	   			}

			}

			if (verboseflag) {

				 fprintf(stderr,"DEBUG: running chpw_nopam() for user %s\n",name);
				 fprintf(stderr,"DEBUG: mode = %d\n",mode);

			}

	   		i=chpw_nopam(pw,pass,mode);

			if (i==-3) {

				fprintf(stderr,"\n%s: failed to create lock %s",prgname,TMPLOCK);


			}

			if (verboseflag) {
				fprintf(stderr,"DEBUG: chpw_nopam() return code = %d \n",i);
				fprintf(stderr,"DEBUG: str_passerror = \"%s\"\n",str_passerror(i));
			}

#ifdef _WITHPAM
   		}
#endif

/*
		if (verboseflag) fprintf(stderr,"DEBUG: running chpw() (_WITHPAM is disabled) for user %s\n",name);
   		i=chpw(pw,pass);
*/


   		if (i!=PASS_SUCCESS) {

      			/* chpw() failed */
      			fprintf(stderr,"\n%s: Can't update password for user %s on line %d",prgname,name,linecounter+1);
#ifdef _WITHPAM

      			if (pamflag) {

      				fprintf(stderr,"\n%s: pam return code = %d\n",pam_strerror(pw->pamh,i),i);
				fprintf(stderr,"\n errno = %d %s\n",errno,strerror(errno));

      			}
#endif

			error=1;

   		}
		else {

			pass_updated++;

		}

	}

   	/* password updated */

	if (error) {

		fprintf(stderr,"\n%s: Sorry, couldn't update all passwords\n",prgname);

	}
   
   	fprintf(stderr,"%s: %d of %d passwords updated...\n\n",prgname,pass_updated,number_of_lines);
   	exit(error);

}
